### PR TITLE
[Feature] Support tokenize function

### LINKED ISSUE #45145: Support tokenize function
- [x] https://github.com/StarRocks/starrocks/pull/58965
## Why I'm doing:
StarRocks has implemented GIN (Generalized Inverted Index), which works by tokenizing fields into individual tokens and building a dictionary out of them. This allows users to perform different semantic searches on this dictionary. However, due to the presence of various tokenizers, the results of tokenization can differ, making it not very intuitive for users to understand how the original field text is tokenized into specific tokens.

## What I'm doing:
Support a tokenize function, to allow users to work with specific tokenizer and get results of tokenization easily.

#### Function definition
```
function tokenize(tokenizer_name: string, content: string) -> list of strings
```
#### Input and output
```
// input 
tokenizer_name: needs to be limited to the existing tokenizers. For now, only support chinese, english, standard.
content: text, but notice that the language corresponding to the text content only achieves the expected effect when work with the specified tokenizer.

// output 
tokens: splited and analyzed by tokenizer
```

#### Example
```
// tokenize with english
mysql> SELECT tokenize('english', 'Today is saturday');
+------------------------------------------+
| tokenize('english', 'Today is saturday') |
+------------------------------------------+
| ["today","is","saturday"]                |
+------------------------------------------+
1 row in set (0.00 sec)

// count word frequency
mysql> select unnest, count(*) as count 
mysql> from t_tokenized_table, unnest(tokenize('english', english_text)) as unnest
mysql> group by unnest order by count;
+----------+-------+
| unnest   | count |
+----------+-------+
| world    |     1 |
| comes    |     1 |
| tap      |     1 |
| the      |     1 |
| from     |     1 |
| sea      |     1 |
| shanghai |     1 |
| water    |     1 |
| hello    |     2 |
+----------+-------+
9 rows in set (0.06 sec)
```

#### Notice
This function can __work independently__ without building GIN for a column, however, it is not advisable to invoke this function to tokenize and construct a dictionary on the massive data during the query time due to the poor performance. In fact, there is no need for the user to explicitly call the tokenize function to build a dictionary at the time of writing. Since both actions have the same behavior, this function is more suitable to be used to troubleshoot some search results that are difficult to understand.

### PR DESCRIPTION
**Pull request by hongkunxu**
## Why I'm doing:
The different results of tokenization provided by various tokenizers are too vague to users, so we need a tokenize function to allow users to figure it out easily.

We are currently researching inverted indexes, so we need a tokenize function to obtain more detailed tokenization results from the parser. I noticed that the original author, dujijun007, is no longer actively maintaining the PR or following up with the StarRocks community. Therefore, I copied the original PR and made some modifications based on the community’s feedback.

## What I'm doing:
Support a tokenize function, like tokenize(<tokenizer_name>, <content>)

Here is the function describe doc: https://docs.google.com/document/d/1oOegC8Qwud0_z08tgytx8KyZB3lwRZwfQCmO1kMw784/edit?usp=sharing

Fixes https://github.com/StarRocks/starrocks/issues/45145 https://github.com/StarRocks/starrocks/issues/61827

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1

### CODE REVIEW

File: be/src/exprs/gin_functions.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <CLucene.h>
```
Comment (alvin-celerdata) [inline be/src/exprs/gin_functions.h:17]: This header is not used in this file; please move it into the CPP files. We should keep the header files included as few as possible for better compilation speed and reduced exposure of third-party code.

File: gensrc/script/functions.py [hunk 2]
```diff
@@ -1398,4 +1398,7 @@
     # data synopse functions
     [181000, 'bar', True, False, 'VARCHAR', ['BIGINT', "BIGINT", 'BIGINT', 'BIGINT'], 'UtilityFunctions::bar'],
     [181001, 'equiwidth_bucket', True, False, 'BIGINT', ['BIGINT', 'BIGINT', 'BIGINT', 'BIGINT'], 'UtilityFunctions::equiwidth_bucket'],
+
+    # gin functions
+    [190000, 'tokenize', True, False, 'ARRAY_VARCHAR', ['VARCHAR', 'VARCHAR'], 'GinFunctions::tokenize', 'GinFunctions::tokenize_prepare', 'GinFunctions::tokenize_close']
```
Comment (alvin-celerdata) [inline gensrc/script/functions.py:1403]: could write an English document to describe the function behavior?

**Review by alvin-phoenix-ai — COMMENTED** [review]
Please change the title from "Enhancement" to "Feature", because this PR will introduce a new function.

File: gensrc/script/functions.py (hunk 2 above)
Comment (INNOCENT-BOY) [reply gensrc/script/functions.py:1403]: I have write english google doc and put it in the "What I am doing" section.

File: be/src/exprs/gin_functions.h (hunk 1 above)
Comment (INNOCENT-BOY) [reply be/src/exprs/gin_functions.h:17]: remove it.

File: gensrc/script/functions.py (hunk 2 above)
Comment (alvin-celerdata) [reply gensrc/script/functions.py:1403]: What I mean is to write a doc in the directory of "/docs/en/sql-reference/sql-functions". Once the feature is released, users can learn about these functions on the StarRocks doc website. Otherwise, no one will know StarRocks has this function.

File: test/sql/test_inverted_index/T/test_tokenize [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+"replicated_storage" = "false",
+"compression" = "LZ4"
+);
+
+INSERT INTO t_tokenized_table VALUES
+(1, 'hello', 'Привет', '你好'),
+(2, 'hello world', 'Привет, мир', '你好世界'),
+(3, 'Shanghai tap water comes from the sea', 'Водопроводная вода в Шанхае поступает из моря', '上海自来水来自海上');
+
+select id, tokenize('english', english_text), tokenize('standard', standard_text), tokenize('chinese', chinese_text) from t_tokenized_table order by id;
```
Comment (alvin-celerdata) [inline test/sql/test_inverted_index/T/test_tokenize:26]: More tests should be added to guarantee the correct behavior, such as the parameter of "Null", empty strings.

Comment (INNOCENT-BOY) [reply test/sql/test_inverted_index/T/test_tokenize:30]: I have added more tests.

File: gensrc/script/functions.py (hunk 2 above)
Comment (INNOCENT-BOY) [reply gensrc/script/functions.py:1403]: Have write a tokenize.md in the sql-functions.

File: test/sql/test_inverted_index/T/test_tokenize (hunk 3 above)
Comment (INNOCENT-BOY) [reply test/sql/test_inverted_index/T/test_tokenize:30]: @alvin-celerdata Please help review again.

**Review by Seaven — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

File: be/src/exprs/gin_functions.cpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    if (scope != FunctionContext::THREAD_LOCAL) {
+        return Status::OK();
+    }
+
+    auto column = context->get_constant_column(0);
+    auto method = ColumnHelper::get_const_value<TYPE_VARCHAR>(column);
+
+    lucene::analysis::Analyzer* analyzer;
+
+    if (method == "english") {
```
Comment (apatrida) [inline be/src/exprs/gin_functions.cpp:39]: This is not an English tokenizer at all, completely wrong tokenization for "english"  

This will break English words in bad ways, doesn't stem, doesn't lem, doesn't deplural, doesn't handle possessives, doesn't handle conjunctions, etc.

completely wrong.

File: be/src/exprs/gin_functions.cpp [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        analyzer = _CLNEW lucene::analysis::SimpleAnalyzer();
+    } else if (method == "standard") {
+        analyzer = _CLNEW lucene::analysis::standard::StandardAnalyzer();
```
Comment (apatrida) [inline be/src/exprs/gin_functions.cpp:42]: This uses "english" stop words by default and is not documented that stop word removal happens, but only for English.  It should also be documented that this uses the Unicode standard for word separations which might alter some test such as possessives.

This is accidentally more English than the English analyzer above.
